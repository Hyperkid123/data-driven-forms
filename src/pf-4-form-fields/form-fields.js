import React from 'react';

import {
  TextInput,
  FormGroup,
  TextContent,
  Text,
  TextVariants,
  TextArea,
  Select,
  SelectOption,
  Checkbox,
  Radio,
  GridItem,
} from '@patternfly/react-core';
import PropTypes from 'prop-types';
import { Field } from 'react-final-form';
import { components } from '../constants';
import { __ } from '../global-functions';
import Condition from '../shared-components/condition';
import FormGroupWrapper from '../shared-components/form-group-wrapper';
import { composeValidators } from '../helpers';
import MultipleChoiceList from './multiple-choice-list';
import NestedForm from '../shared-components/sub-form';

const selectComponent = ({ componentType, input, options, isReadOnly, isDisabled, ...rest }) => ({
  [components.TEXT_FIELD]: () => (
    <TextInput
      { ...input }
      { ...rest }
      isReadOnly={ isReadOnly }
      isDisabled={ isDisabled }
    />
  ),
  [components.TEXTAREA_FIELD]: () => <TextArea { ...input } { ...rest } />,
  [components.SELECT_COMPONENT]: () => (
    <Select { ...input } { ...rest }>
      { options.map(props => (<SelectOption key={ props.value || props.label } { ...props } label={ props.label.toString() }/>)) }
    </Select>
  ),
  [components.CHECKBOX]: () =>  <Checkbox { ...input } { ...rest }/>,
  [components.RADIO]: () => (
    <Radio
      { ...input }
      { ...rest }
      value={ !!input.value }
      onChange={ () => input.onChange(input.value) }
    />
  ),
})[componentType];

const FinalFormField = ({
  componentType,
  label,
  isRequired,
  helperText,
  meta,
  description,
  ...rest
}) => {
  const { error, touched } = meta;
  const showError = touched && error;

  return (
    <GridItem span={ 12 }>
      <FormGroup
        isRequired={ isRequired }
        label={ __(label) }
        fieldId={ rest.id }
        isValid={ !showError }
        helperText={ __(helperText) }
        helperTextInvalid={ __(meta.error) }
      >
        { description && <TextContent><Text component={ TextVariants.small }>{ __(description) }</Text></TextContent> }
        { selectComponent({ componentType, ...rest, isValid: !showError })() }
      </FormGroup>
    </GridItem>
  );
};

FinalFormField.propTypes = {
  componentType: PropTypes.string.isRequired,
  label: PropTypes.string,
  isRequired: PropTypes.bool,
  helperText: PropTypes.string,
  meta: PropTypes.object.isRequired,
  description: PropTypes.string,
};

FinalFormField.defaultProps = {
  isRequired: false,
  description: undefined,
};

const SingleChoiceList = ({ validate, ...props }) => (
  <Field
    { ...props }
    validate={ composeValidators(...props.validate || []) }
    render={ ({ input, meta, options, label, isRequired, helperText, ...rest }) => {
      const { error, touched } = meta;
      const showError = touched && error;
      return (
        <FormGroup
          label={ __(label) }
          isRequired={ isRequired }
          fieldId={ rest.id }
          helperText={ __(helperText) }
          helperTextInvalid= { __(meta.error) }
          isValid={ !showError }
        >
          { options.map(option => (
            <Field
              id={ `${rest.id}-${option.value}` }
              key={ `${props.name}-${option.value}` }
              name={ props.name }
              label={ __(option.label) }
              value={ option.value }
              aria-label={ option['aria-label'] || option.label }
              type="radio"
              componentType={ components.RADIO }
              render={ props => selectComponent(props)() }
            />
          )) }
        </FormGroup>
      );} } />);

const SingleCheckboxField = ({ isReadOnly, helperText, ...props }) => (
  <Field
    { ...props }
    validate={ composeValidators(...props.validate || []) }
    type="checkbox"
    render={ rest => selectComponent({ ...rest })() }
  />
);

const CheckboxGroupField = ({ options, ...rest }) =>
  options ? <MultipleChoiceList options={ options } { ...rest }/>
    : (
      <SingleCheckboxField
        aria-label={ rest['aria-label'] || rest.label }
        { ...rest }
        componentType={ components.CHECKBOX }
      />
    );

const fieldMapper = type => ({
  [components.TEXT_FIELD]: props => <FormGroupWrapper { ...props } formField={ FinalFormField } />,
  [components.SELECT_COMPONENT]: props =>
    <FormGroupWrapper componentType={ components.SELECT_COMPONENT } formField={ FinalFormField } { ...props } />,
  [components.TEXTAREA_FIELD]: props => <FormGroupWrapper componentType={ components.TEXTAREA_FIELD } formField={ FinalFormField } { ...props } />,
  [components.CHECKBOX]: props => <CheckboxGroupField { ...props }/>,
  [components.RADIO]: props => <SingleChoiceList { ...props }/>,
  [components.SUB_FORM]: props => <NestedForm { ...props } />,
})[type];

const FormConditionWrapper = ({ condition, children }) => (condition ? (
  <Condition { ...condition }>
    { children }
  </Condition>
) : children);

const FieldInterface = ({
  meta,
  dataType,
  condition,
  validate,
  componentType,
  initialKey,
  ...props
}) => (
  <FormConditionWrapper condition={ condition }>
    { fieldMapper(componentType)({
      ...props,
      validate: Array.isArray(validate) ? validate : [ validate ],
      componentType,
      id: props.id || props.name,
    }) }
  </FormConditionWrapper>
);

FieldInterface.propTypes = {
  meta: PropTypes.object,
  condition: PropTypes.shape({
    when: PropTypes.string.isRequired,
    is: PropTypes.oneOfType([ PropTypes.array, PropTypes.string ]).isRequired,
  }),
  validate: PropTypes.oneOfType([ PropTypes.array, PropTypes.func ]),
  componentType: PropTypes.oneOf([
    components.RADIO,
    components.CHECKBOX,
    components.SELECT_COMPONENT,
    components.TEXTAREA_FIELD,
    components.TEXT_FIELD,
    components.SUB_FORM,
  ]).isRequired,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  dataType: PropTypes.any,
  initialKey: PropTypes.any,
};

export const TextField = props => <FieldInterface { ...props } componentType={ components.TEXT_FIELD }  />;
export const TextareaField = props => <FieldInterface { ...props }  componentType={ components.TEXTAREA_FIELD } />;
export const SelectField = props=> <FieldInterface { ...props } componentType={ components.SELECT_COMPONENT } />;
export const CheckboxGroup = props => <FieldInterface { ...props } componentType={ components.CHECKBOX }/>;
export const SubForm = props => <FieldInterface { ...props } componentType={ components.SUB_FORM } />;
export const RadioGroup = props => <FieldInterface { ...props } componentType={ components.RADIO } />;
