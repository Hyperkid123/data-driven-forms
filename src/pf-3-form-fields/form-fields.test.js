import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Form, Field } from 'react-final-form';
import { FormGroup } from 'patternfly-react';
import { TextField, TextareaField, SelectField, RadioGroup, CheckboxGroup, SubForm } from './form-fields';
import ComponentType, { configureContext } from '../renderer-context';
import MultipleChoiceList from './multiple-choice-list';

describe('Pf3 form fields', () => {
  const FormWrapper = ({ children, ...props }) => (
    <Form onSubmit={ jest.fn() } { ...props }>
      { () => (
        <ComponentType.Provider value={ configureContext({ formType: 'pf3' }) }>
          { children }
        </ComponentType.Provider>
      ) }
    </Form>
  );

  describe('<TextField />', () => {
    let initialProps;
    beforeEach(() => {
      initialProps = {
        name: 'text-field',
        type: 'text',
      };
    });

    it('should render correctly in valid state', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextField { ...initialProps }/>
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });

    it('should render correctly in error state', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextField { ...initialProps }/>
        </FormWrapper>
      );
      wrapper.find(Field).first().instance().setState(prevState => ({
        state: {
          ...prevState.state,
          error: 'Has some error',
          touched: true,
        },
      }));
      wrapper.find(Form).first().update();
      expect(toJson(wrapper.find(FormGroup).first())).toMatchSnapshot();
    });

    it('should render correctly in valid state with helper text and description', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextField { ...initialProps } helperText="i am helper text" description="i am description" />
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });
  });

  describe('<TextareaField />', () => {
    let initialProps;
    beforeEach(() => {
      initialProps = {
        name: 'textarea-field',
      };
    });

    it('should render correctly in valid state', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextareaField { ...initialProps }/>
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });

    it('should render correctly in error state', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextareaField { ...initialProps }/>
        </FormWrapper>
      );
      wrapper.find(Field).first().instance().setState(prevState => ({
        state: {
          ...prevState.state,
          error: 'Has some error',
          touched: true,
        },
      }));
      wrapper.find(Form).first().update();
      expect(toJson(wrapper.find(FormGroup).first())).toMatchSnapshot();
    });

    it('should render correctly in valid state with helper text and description', () => {
      const wrapper = mount(
        <FormWrapper>
          <TextareaField { ...initialProps } helperText="i am helper text" description="i am description" />
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });
  });

  describe('<SelectField />', () => {
    let initialProps;
    beforeEach(() => {
      initialProps = {
        name: 'select-field',
        options: [{
          label: 'foo',
          value: 'foo',
        }, {
          label: 'bar',
          value: 'bar',
        }],
      };
    });

    it('should render correctly in valid state', () => {
      const wrapper = mount(
        <FormWrapper>
          <SelectField { ...initialProps }/>
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });

    it('should render correctly in error state', () => {
      const wrapper = mount(
        <FormWrapper>
          <SelectField { ...initialProps }/>
        </FormWrapper>
      );
      wrapper.find(Field).first().instance().setState(prevState => ({
        state: {
          ...prevState.state,
          error: 'Has some error',
          touched: true,
        },
      }));
      wrapper.find(Form).first().update();
      expect(toJson(wrapper.find(FormGroup).first())).toMatchSnapshot();
    });

    it('should render correctly in valid state with helper text and description', () => {
      const wrapper = mount(
        <FormWrapper>
          <SelectField { ...initialProps } helperText="i am helper text" description="i am description" />
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });
  });

  describe('<RadioGroup />', () => {
    let initialProps;
    beforeEach(() => {
      initialProps = {
        name: 'radio-field',
        options: [{
          label: 'foo',
          value: 'foo',
        }, {
          label: 'bar',
          value: 'bar',
        }],
      };
    });

    it('should render correctly', () => {
      const wrapper = mount(
        <FormWrapper>
          <RadioGroup { ...initialProps }/>
        </FormWrapper>
      );
      expect(toJson(wrapper.find(FormGroup))).toMatchSnapshot();
    });
  });

  describe('<CheckboxGroup />', () => {
    let initialProps;
    beforeEach(() => {
      initialProps = {
        name: 'checkbox-field',
        options: [{
          label: 'foo',
          value: 'foo',
        }, {
          label: 'bar',
          value: 'bar',
        }],
      };
    });

    it('should render multiple choice list', () => {
      const wrapper = mount(
        <FormWrapper>
          <CheckboxGroup { ...initialProps }/>
        </FormWrapper>
      );
      expect(wrapper.find(MultipleChoiceList)).toBeTruthy();
    });

    it('should render single checkbox', () => {
      const wrapper = mount(
        <FormWrapper>
          <CheckboxGroup { ...initialProps } options={ undefined }/>
        </FormWrapper>
      );
      expect(wrapper.find(MultipleChoiceList)).toBeTruthy();
    });
  });

  describe('<SubForm />', () => {
    it('should render subForm component with TextField', () => {
      const wrapper = mount(
        <FormWrapper>
          <SubForm
            fields={ [{
              component: 'text-field',
              name: 'nested-field',
            }] }
            name="sub-form"
            title="SubForm title"
            description="SubForm description"
            renderForm={ () => <div id="nested-item">Foo</div> }
          />
        </FormWrapper>
      );
      expect(toJson(wrapper.find(SubForm))).toMatchSnapshot();
    });
  });
});
