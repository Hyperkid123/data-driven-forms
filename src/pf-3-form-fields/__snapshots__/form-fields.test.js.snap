// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pf3 form fields <RadioGroup /> should render correctly 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <Field
          format={[Function]}
          key="radio-field-foo"
          name="radio-field"
          parse={[Function]}
          render={[Function]}
          type="radio"
          value={true}
        >
          <Radio
            bsClass="radio"
            checked={false}
            disabled={false}
            inline={false}
            name="radio-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            title=""
            value={true}
          >
            <div
              className="radio"
            >
              <label
                title=""
              >
                <input
                  checked={false}
                  disabled={false}
                  name="radio-field"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value={true}
                />
                foo
              </label>
            </div>
          </Radio>
        </Field>
        <Field
          format={[Function]}
          key="radio-field-bar"
          name="radio-field"
          parse={[Function]}
          render={[Function]}
          type="radio"
          value={true}
        >
          <Radio
            bsClass="radio"
            checked={false}
            disabled={false}
            inline={false}
            name="radio-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            title=""
            value={true}
          >
            <div
              className="radio"
            >
              <label
                title=""
              >
                <input
                  checked={false}
                  disabled={false}
                  name="radio-field"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value={true}
                />
                bar
              </label>
            </div>
          </Radio>
        </Field>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <SelectField /> should render correctly in error state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState="error"
>
  <div
    className="form-group has-error"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <Select
          arrowRenderer={[Function]}
          autosize={true}
          backspaceRemoves={true}
          backspaceToRemoveMessage="Press backspace to remove {label}"
          className="has-error final-form-select"
          clearAllText="Clear all"
          clearRenderer={[Function]}
          clearValueText="Clear value"
          clearable={true}
          closeOnSelect={true}
          deleteRemoves={true}
          delimiter=","
          disabled={false}
          escapeClearsValue={true}
          filterOptions={[Function]}
          id="select-field"
          ignoreAccents={true}
          ignoreCase={true}
          inputProps={Object {}}
          isLoading={false}
          joinValues={false}
          labelKey="label"
          matchPos="any"
          matchProp="any"
          menuBuffer={0}
          menuRenderer={[Function]}
          multi={false}
          name="select-field"
          noResultsText="No results found"
          onBlur={[Function]}
          onBlurResetsInput={true}
          onChange={[Function]}
          onCloseResetsInput={true}
          onFocus={[Function]}
          onSelectResetsInput={true}
          openOnClick={true}
          optionClassName="final-form-select-option"
          optionComponent={[Function]}
          options={
            Array [
              Object {
                "label": "foo",
                "value": "foo",
              },
              Object {
                "label": "bar",
                "value": "bar",
              },
            ]
          }
          pageSize={5}
          placeholder="Please choose"
          removeSelected={true}
          required={false}
          rtl={false}
          scrollMenuIntoView={true}
          searchable={false}
          simpleValue={false}
          tabSelectsValue={true}
          trimFilter={true}
          value=""
          valueComponent={[Function]}
          valueKey="value"
        >
          <div
            className="Select has-error final-form-select is-clearable Select--single"
          >
            <div
              className="Select-control"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="Select-multi-value-wrapper"
                id="react-select-3--value"
              >
                <div
                  className="Select-placeholder"
                >
                  Please choose
                </div>
                <div
                  aria-activedescendant="react-select-3--value"
                  aria-disabled="false"
                  aria-expanded={false}
                  aria-owns=""
                  className="Select-input"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  style={
                    Object {
                      "border": 0,
                      "display": "inline-block",
                      "width": 1,
                    }
                  }
                  tabIndex={0}
                />
              </div>
              <span
                className="Select-arrow-zone"
                onMouseDown={[Function]}
              >
                <span
                  className="Select-arrow"
                  onMouseDown={[Function]}
                />
              </span>
            </div>
          </div>
        </Select>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            Has some error
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <SelectField /> should render correctly in valid state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <Select
          arrowRenderer={[Function]}
          autosize={true}
          backspaceRemoves={true}
          backspaceToRemoveMessage="Press backspace to remove {label}"
          className=" final-form-select"
          clearAllText="Clear all"
          clearRenderer={[Function]}
          clearValueText="Clear value"
          clearable={true}
          closeOnSelect={true}
          deleteRemoves={true}
          delimiter=","
          disabled={false}
          escapeClearsValue={true}
          filterOptions={[Function]}
          id="select-field"
          ignoreAccents={true}
          ignoreCase={true}
          inputProps={Object {}}
          isLoading={false}
          joinValues={false}
          labelKey="label"
          matchPos="any"
          matchProp="any"
          menuBuffer={0}
          menuRenderer={[Function]}
          multi={false}
          name="select-field"
          noResultsText="No results found"
          onBlur={[Function]}
          onBlurResetsInput={true}
          onChange={[Function]}
          onCloseResetsInput={true}
          onFocus={[Function]}
          onSelectResetsInput={true}
          openOnClick={true}
          optionClassName="final-form-select-option"
          optionComponent={[Function]}
          options={
            Array [
              Object {
                "label": "foo",
                "value": "foo",
              },
              Object {
                "label": "bar",
                "value": "bar",
              },
            ]
          }
          pageSize={5}
          placeholder="Please choose"
          removeSelected={true}
          required={false}
          rtl={false}
          scrollMenuIntoView={true}
          searchable={false}
          simpleValue={false}
          tabSelectsValue={true}
          trimFilter={true}
          value=""
          valueComponent={[Function]}
          valueKey="value"
        >
          <div
            className="Select  final-form-select is-clearable Select--single"
          >
            <div
              className="Select-control"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="Select-multi-value-wrapper"
                id="react-select-2--value"
              >
                <div
                  className="Select-placeholder"
                >
                  Please choose
                </div>
                <div
                  aria-activedescendant="react-select-2--value"
                  aria-disabled="false"
                  aria-expanded={false}
                  aria-owns=""
                  className="Select-input"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  style={
                    Object {
                      "border": 0,
                      "display": "inline-block",
                      "width": 1,
                    }
                  }
                  tabIndex={0}
                />
              </div>
              <span
                className="Select-arrow-zone"
                onMouseDown={[Function]}
              >
                <span
                  className="Select-arrow"
                  onMouseDown={[Function]}
                />
              </span>
            </div>
          </div>
        </Select>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <SelectField /> should render correctly in valid state with helper text and description 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <Select
          arrowRenderer={[Function]}
          autosize={true}
          backspaceRemoves={true}
          backspaceToRemoveMessage="Press backspace to remove {label}"
          className=" final-form-select"
          clearAllText="Clear all"
          clearRenderer={[Function]}
          clearValueText="Clear value"
          clearable={true}
          closeOnSelect={true}
          deleteRemoves={true}
          delimiter=","
          disabled={false}
          escapeClearsValue={true}
          filterOptions={[Function]}
          id="select-field"
          ignoreAccents={true}
          ignoreCase={true}
          inputProps={Object {}}
          isLoading={false}
          joinValues={false}
          labelKey="label"
          matchPos="any"
          matchProp="any"
          menuBuffer={0}
          menuRenderer={[Function]}
          multi={false}
          name="select-field"
          noResultsText="No results found"
          onBlur={[Function]}
          onBlurResetsInput={true}
          onChange={[Function]}
          onCloseResetsInput={true}
          onFocus={[Function]}
          onSelectResetsInput={true}
          openOnClick={true}
          optionClassName="final-form-select-option"
          optionComponent={[Function]}
          options={
            Array [
              Object {
                "label": "foo",
                "value": "foo",
              },
              Object {
                "label": "bar",
                "value": "bar",
              },
            ]
          }
          pageSize={5}
          placeholder="Please choose"
          removeSelected={true}
          required={false}
          rtl={false}
          scrollMenuIntoView={true}
          searchable={false}
          simpleValue={false}
          tabSelectsValue={true}
          trimFilter={true}
          value=""
          valueComponent={[Function]}
          valueKey="value"
        >
          <div
            className="Select  final-form-select is-clearable Select--single"
          >
            <div
              className="Select-control"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="Select-multi-value-wrapper"
                id="react-select-4--value"
              >
                <div
                  className="Select-placeholder"
                >
                  Please choose
                </div>
                <div
                  aria-activedescendant="react-select-4--value"
                  aria-disabled="false"
                  aria-expanded={false}
                  aria-owns=""
                  className="Select-input"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  style={
                    Object {
                      "border": 0,
                      "display": "inline-block",
                      "width": 1,
                    }
                  }
                  tabIndex={0}
                />
              </div>
              <span
                className="Select-arrow-zone"
                onMouseDown={[Function]}
              >
                <span
                  className="Select-arrow"
                  onMouseDown={[Function]}
                />
              </span>
            </div>
          </div>
        </Select>
        <HelpBlock
          bsClass="help-block"
          style={
            Object {
              "color": "#767676",
            }
          }
        >
          <span
            className="help-block"
            style={
              Object {
                "color": "#767676",
              }
            }
          >
            i am description
          </span>
        </HelpBlock>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            i am helper text
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <SubForm /> should render subForm component with TextField 1`] = `
<SubForm
  description="SubForm description"
  fields={
    Array [
      Object {
        "component": "text-field",
        "name": "nested-field",
      },
    ]
  }
  name="sub-form"
  renderForm={[Function]}
  title="SubForm title"
>
  <FieldInterface
    componentType="sub-form"
    description="SubForm description"
    fields={
      Array [
        Object {
          "component": "text-field",
          "name": "nested-field",
        },
      ]
    }
    name="sub-form"
    renderForm={[Function]}
    title="SubForm title"
  >
    <FormConditionWrapper>
      <SubForm
        componentType="sub-form"
        description="SubForm description"
        fields={
          Array [
            Object {
              "component": "text-field",
              "name": "nested-field",
            },
          ]
        }
        id="sub-form"
        name="sub-form"
        renderForm={[Function]}
        title="SubForm title"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <Col
          bsClass="col"
          componentClass="div"
          xs={12}
        >
          <div
            className="col-xs-12"
          >
            <h3>
              SubForm title
            </h3>
            <p>
              SubForm description
            </p>
            <div
              id="nested-item"
            >
              Foo
            </div>
          </div>
        </Col>
      </SubForm>
    </FormConditionWrapper>
  </FieldInterface>
</SubForm>
`;

exports[`Pf3 form fields <TextField /> should render correctly in error state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState="error"
>
  <div
    className="form-group has-error"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          id="text-field"
          name="text-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        >
          <input
            className="form-control"
            id="text-field"
            name="text-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </FormControl>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            Has some error
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <TextField /> should render correctly in valid state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          id="text-field"
          name="text-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        >
          <input
            className="form-control"
            id="text-field"
            name="text-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </FormControl>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <TextField /> should render correctly in valid state with helper text and description 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          id="text-field"
          name="text-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        >
          <input
            className="form-control"
            id="text-field"
            name="text-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </FormControl>
        <HelpBlock
          bsClass="help-block"
          style={
            Object {
              "color": "#767676",
            }
          }
        >
          <span
            className="help-block"
            style={
              Object {
                "color": "#767676",
              }
            }
          >
            i am description
          </span>
        </HelpBlock>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            i am helper text
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <TextareaField /> should render correctly in error state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState="error"
>
  <div
    className="form-group has-error"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="textarea"
          id="textarea-field"
          name="textarea-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value=""
        >
          <textarea
            className="form-control"
            id="textarea-field"
            name="textarea-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value=""
          />
        </FormControl>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            Has some error
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <TextareaField /> should render correctly in valid state 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="textarea"
          id="textarea-field"
          name="textarea-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value=""
        >
          <textarea
            className="form-control"
            id="textarea-field"
            name="textarea-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value=""
          />
        </FormControl>
      </div>
    </Col>
  </div>
</FormGroup>
`;

exports[`Pf3 form fields <TextareaField /> should render correctly in valid state with helper text and description 1`] = `
<FormGroup
  bsClass="form-group"
  validationState={null}
>
  <div
    className="form-group"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <FormControl
          bsClass="form-control"
          componentClass="textarea"
          id="textarea-field"
          name="textarea-field"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          value=""
        >
          <textarea
            className="form-control"
            id="textarea-field"
            name="textarea-field"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value=""
          />
        </FormControl>
        <HelpBlock
          bsClass="help-block"
          style={
            Object {
              "color": "#767676",
            }
          }
        >
          <span
            className="help-block"
            style={
              Object {
                "color": "#767676",
              }
            }
          >
            i am description
          </span>
        </HelpBlock>
        <HelpBlock
          bsClass="help-block"
        >
          <span
            className="help-block"
          >
            i am helper text
          </span>
        </HelpBlock>
      </div>
    </Col>
  </div>
</FormGroup>
`;
